fabric/dataclient.go

```diff
@@ -397,36 +452,146 @@ func convertOne(fg *Fabric) ([]*eskip.Route, error) {
+				Name: traffic.ServicePort.StrVal,
+				Port: traffic.ServicePort.IntValue(),
+			})
+			// TODO(sszuecs): maybe check that endpoints are not 0, but what if all of them are 0. Maybe better to shortcute the routes with `status(502) -> <shunt>` in this case.
+
+			for _, host := range esp.Hosts {
+				log.Debugf("x-external-service-provider host=%s svc=%s portName=%s, portNumber=%d", host, traffic.ServiceName, traffic.ServicePort.StrVal, traffic.ServicePort.IntValue())
+
+				routes = append(routes, createRoutes(fg, globalHostRouteDone, trafficParam, noopCount, ridSuffix, host, lbAlgorithm, endpoints, adminArgs, allowedOrigins)...)
+			}
+			globalHostRouteDone = true
+			noopCount--
+		}
+
+	}
+
+	// x-fabric-service
 	for _, fabsvc := range fg.Spec.Service {
 		host := fabsvc.Host
 
 		// TODO(sszuecs): cleanup this hack and think about ingress v1, do we want to change svc def in Fabric?
 		svcName, svcPortName, svcPortNumber := getKubeSvc(fabsvc)
 		// TODO(sszuecs): fix how to get endpoints
-		endpoints := getEndpointsByService(fg.Metadata.Namespace, fg.Metadata.Name, "tcp", &servicePort{
+		endpoints := getEndpointsByService(fg.Metadata.Namespace, svcName, "tcp", &servicePort{
 			Name: svcPortName,
 			Port: svcPortNumber,
 		})
 		log.Debugf("fabsvc host=%s svc=%s portName=%s, portNumber=%d", host, svcName, svcPortName, svcPortNumber)
 
-		be := ""
-		var bt eskip.BackendType = eskip.LBBackend
-		if len(endpoints) == 1 {
-			be = endpoints[0]
-			bt = eskip.NetworkBackend
-			endpoints = nil
-		}
-		eskipBackend := &eskipBackend{
-			Type:        bt,
-			backend:     be,
-			lbAlgorithm: lbAlgorithm,
-			lbEndpoints: endpoints,
+		routes = append(routes, createRoutes(fg, false, -1, -1, "", host, lbAlgorithm, endpoints, adminArgs, allowedOrigins)...)
+
+	}
+
+	// TODO(sszuecs): make sure errors are reported
+	// fmt.Errorf("failed to convert fabricgateway %s/%s: %v", fg.Metadata.Namespace, fg.Metadata.Name, err)
+
+	fd, err := os.Create("/tmp/foo.eskip")
+	if err != nil {
+		return nil, err
+	}
+	defer fd.Close()
+
+	eskip.Fprint(fd, eskip.PrettyPrintInfo{Pretty: true, IndentStr: "\t"}, routes...)
+
+	return routes, nil
+}
+
+// calculateTraffic returns parameters that can be feed into Traffic()
+// predicates and the max count of True() predicates to be used and to
+// be decreased by the user of this function.
+func calculateTraffic(trs []*zv1.ActualTraffic) (map[string]float64, int) {
+	trafficMap := make(map[string]float64)
+	noopCount := 0
+	var weightsSum float64
+	for _, traffic := range trs {
+		if traffic.Weight <= 0 {
+			continue
 		}
+		trafficMap[traffic.ServiceName] = traffic.Weight
+		weightsSum += traffic.Weight
+		noopCount += 1
+	}
+	noopCount -= 2 // 1 route has no Traffic(), 1 route has only Traffic(), and rest needs True()s
+	if noopCount < 0 {
+		noopCount = 0
+	}
 
-		defaultScopePrivileges := []interface{}{
-			"uid",
+	keys := make([]string, 0, len(trafficMap))
```

Comment:
Can be replaced with our utility func: https://github.com/szuecs/skp-ctrl-pln/blob/main/fabric/util.go#L15

Reply:
No we would need to create another one, signature would miss match and I also tried it, but no: `trafficMap := make(map[string]float64)` 
I hope this can be fixed with Go 1.18 and generics